relay-replays/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+//! (segment 0 and 1) have completed ingestion.
+//!
+//! # Protocol
+//!
+//! Relay is expecting a JSON object with some mandatory metadata.  However, environment and user
+//! metadata is usually sent in addition to the minimal payload.
+//!
+//! ```json
+//! {
+//!     "type": "replay_event",
+//!     "replay_id": "d2132d31b39445f1938d7e21b6bf0ec4",
+//!     "event_id": "63c5b0f895441a94340183c5f1e74cd4",
+//!     "segment_id": 0,
+//!     "timestamp": 1597976392.6542819,
+//!     "replay_start_timestamp": 1597976392.6542819,
+//!     "urls": ["https://sentry.io"],
+//!     "error_ids": ["d2132d31b39445f1938d7e21b6bf0ec4"],
+//!     "trace_ids": ["63c5b0f895441a94340183c5f1e74cd4"],
+//!     "requests": {
+//!         "headers": {"User-Agent": "Mozilla/5.0..."}
+//!     },
+//! }
+//! ```
+use relay_general::user_agent::{parse_device, parse_os, parse_user_agent, Device};
+use serde::{Deserialize, Serialize};
+use serde_json::Error;
+use std::collections::HashMap;
+use std::fmt::Write;
+use std::net::IpAddr;
+
+pub fn normalize_replay_event(
+    replay_bytes: &[u8],
+    detected_ip_address: Option<IpAddr>,
+) -> Result<Vec<u8>, Error> {
+    let mut replay_input: ReplayInput = serde_json::from_slice(replay_bytes)?;
+
+    // Set user-agent metadata.
+    replay_input.set_user_agent_meta();
+
+    // Set user ip-address if needed.
+    match detected_ip_address {
+        Some(ip_address) => replay_input.set_user_ip_address(ip_address),
+        None => (),
+    }
+
+    serde_json::to_vec(&replay_input)
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct ReplayInput {
+    #[serde(rename = "type")]
+    ty: String,
+    replay_id: String,
+    event_id: String,
+    segment_id: u8,
+    timestamp: f64,
+    replay_start_timestamp: Option<f64>,
+    urls: Vec<String>,
+    error_ids: Vec<String>,
+    trace_ids: Vec<String>,
+    contexts: Option<Contexts>,
+    dist: Option<String>,
+    platform: Option<String>,
+    environment: Option<String>,
+    release: Option<String>,
+    tags: Option<HashMap<String, String>>,
+    user: Option<User>,
+    sdk: Option<VersionedMeta>,
+    requests: Requests,
+}
+
+impl ReplayInput {
+    pub fn set_user_agent_meta(&mut self) {
+        let user_agent = &self.requests.headers.user_agent;
+
+        let device = parse_device(user_agent);
+
+        let ua = parse_user_agent(user_agent);
+        let browser_struct = VersionedMeta {
+            name: ua.family,
+            version: get_version(&ua.major, &ua.minor, &ua.patch),
+        };
+
+        let os = parse_os(user_agent);
+        let os_struct = VersionedMeta {
+            name: os.family,
+            version: get_version(&os.major, &os.minor, &os.patch),
+        };
+
+        self.contexts = Some(Contexts {
+            device: Some(device),
+            browser: Some(browser_struct),
+            os: Some(os_struct),
+        })
+    }
+
+    pub fn set_user_ip_address(&mut self, ip_address: IpAddr) {
+        match &self.user {
```

Comment:
This match performs a redundant replacement and clones values. The entire function can be simplified as:

```rust
let user = self.user.get_or_insert_default();
user.ip_address.get_or_insert_with(|| ip_address.to_string());
```

This assumes that there is `#[derive(Default)]` on `User`, which can be helpful in other cases as well.